### DISCUSSION TITLE
Best way to support tapret in wallet descriptors

### DISCUSSION THREAD

**Discussion by oacdutra** [BP-WG/descriptor-wallet#53] [Ideas]
According to this [comment](https://github.com/BP-WG/descriptor-wallet/blob/master/src/bin/btc-cold.rs#L172), the descriptor change is provided by the same wallet used to provide the inputs. 

I think the command line needs an optional parameter to allow choose your change descriptor. 

This important feature for TapRet commitments, because always we need a TapTree to add commitment.

**TapRet commitment user case (example):**

1. We create a Taptree output descriptor with OP_CHECKSIG.
```
change_desc="tr(A,pk(B))#mjnpzw6a"
```

2. We construct a PSBT file.
```
btc-cold construct --input  <INPUTS> <WALLET_FILE> <PSBT_FILE> <FEE> --allow-tapret-path 1 --account-file <ACCOUNT_FILE> --change $change_desc 
```

3. After `RGB finalize`, the Taptree is updated with the commitment script 

4. If we want to spend the bitcoins of the output descriptor, we need to create a new wallet based on the script of the step 1
```
btc-cold create  $change_desc <WALLET_FILE> --account-file <ACCOUNT_FILE>
```

5. Use the new wallet to construct a new PSBT file.

Does make sense?

**Comment by dr-orlovsky** [comment]
You are right about that existance of tapret commiments makes it harder to deal with change outputs; but I do not agree that this should be addressed by creating a dedicated descriptor. First of all, this will significantly complicate the wallet business logic (the need to represent a wallet with multiple descriptors). Also, we to keep in mind that taprets may be theoretically present not just in the change addresses.

My plan was to support extended descriptor format supporting tapret fragment and a list of tweaks which should be looked for at specific indexes.

I did some attempts in this early prototype concept: https://github.com/BP-WG/stingerjet-docs (see the end of the README, the taprets final section of a sample). But before doing a new wallet descriptor languague we can start with just adding tapret fragment - or using dedicated `tapret` descriptor (instead of `tr`). I think the later is better actually, and I remember miniscript guys were ready to add more descriptors to the library.

So the descriptor may look like this:

`tapret(xpub, {}, tapret1qkfdxhgs8trv6gmxh0xd0yeu8kyx8chtj4zjygx)`, where all descriptor arguments after the second one are just tapret commitments which were created.

**Reply by oacdutra** [reply]
WOW, thanks for the explain these all concepts.

I will start explore miniscript to solve this.
